### ISSUE TITLE
0.19.0 -> 0.19.1 regression: compiler goes into infinite lool

### ISSUE THREAD

**Issue by ilya-klyuchnikov** [elm/compiler#2029]
**Quick Summary:** The well-typed program which is accepted by 0.19.0 results into infinite loop of 0.19.1

## SSCCE

```elm
module Weak exposing (..)

type alias X1 r a = {r | x : a}

type alias X2 t = X1 t t

x2 : X2 {}
x2 = {x = {}}

id : X2 a -> X2 a
id x = x

x3 = id x2
```

elm 0.19.0:

```
elm make src/Weak.elm
Success!
```

elm 0.19.1: loop

- **Elm:** 0.19.1
- **Browser:** Google Chrome 78.0.3904.87
- **Operating System:** MacOS 10.14.6

## Additional Details

Actually, the program is well-typed.

**Comment by nikita-volkov** [comment]
Same happens for the following code

```elm
type alias Lens a b = {
    get : a -> b,
    set : b -> a -> a
  }

type alias State inputState = {
    input : inputState
  }

inputLens : Lens (State inputState) inputState
inputLens = Lens .input (\ a b -> { a | input = b })
```

**Comment by zwilias** [comment]
@ilya-klyuchnikov I cannot reproduce the example in the first post:

![image](https://user-images.githubusercontent.com/1038427/70793855-97f8f380-1d9c-11ea-8999-bdf5ae1ef44c.png)


@nikita-volkov It's usually best to create a separate issue. Your example is not well-typed, btw. `set` should be `b -> a -> a` but you're giving it an `a -> b -> a`. Try `(\b a -> { a | input = b })`.

**Comment by nikita-volkov** [comment]
@zwilias Yes, my code is ill-typed, but it's not the point. The point is that there is a bug in the compiler that throws it into an infinite loop on that code.

**Comment by nikita-volkov** [comment]
I've created a dedicated issue for my code sample: #2036
